<a href="https://colab.research.google.com/github/Maneesh290318/LLM_Creation/blob/main/TransformerLLMCreation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Task
Build and evaluate a custom 3-layer Decoder Transformer model specialized in Indian cultural and geographical knowledge. The project involves synthetic data generation for the Indian context, implementing a neural network architecture with self-attention, training the model, and performing a comprehensive evaluation of its performance including hallucination analysis and technical metrics like perplexity and accuracy.

## Project Overview and Explanation

### Subtask:
Create an introductory cell explaining the project goals, the 3-layer Decoder architecture, and the focus on Indian cultural data.


# Project Overview: Custom 3-Layer Decoder Transformer for Indian Cultural Knowledge

### Introduction
This project aims to design, implement, and evaluate a specialized **Decoder-only Transformer model** with a 3-layer architecture. Unlike general-purpose models, this Transformer is specifically trained on a synthetically generated dataset focused on **Indian culture, geography, and landmarks**.

### Project Goals
*   **Custom Architecture**: Build a lightweight yet effective 3-layer neural network using self-attention mechanisms.
*   **Domain Specificity**: Focus on accurately representing Indian states, capitals, festivals, and geographical facts.
*   **Evaluation Excellence**: Measure performance using technical metrics like Perplexity and Accuracy, while performing a detailed Hallucination Analysis.

### The 3-Layer Architecture Explained
To make this accessible to all stakeholders, we break down our '3-layer' approach as follows:
1.  **Input Layer (Contextualization)**: Where the model receives and encodes the text sequences into numbers.
2.  **Hidden Layer (Processing/Attention)**: The core 'brain' where the model uses self-attention to understand the relationships between different Indian entities (e.g., linking 'Karnataka' to 'Bangalore').
3.  **Output Layer (Prediction)**: Where the model generates the most probable next word in the sequence.

### Why Indian Cultural Data?
By using a targeted synthetic dataset, we ensure high-quality 'ground truth' information. This allows us to strictly evaluate the model's ability to learn specific patterns without the noise found in general web-scraped data.

## Synthetic Data Generation for Indian Context

### Subtask:
Programmatically generate a dataset covering Indian states, capitals, landmarks, and festivals to serve as the ground truth.


**Reasoning**:
I will start by creating the raw data structure and the sentence generation logic using Python.



In [ ]:
import random

# Data structure representing Indian cultural and geographical facts
india_data = [
    {"state": "Karnataka", "capital": "Bangalore", "landmark": "Mysore Palace", "festival": "Mysore Dasara"},
    {"state": "Maharashtra", "capital": "Mumbai", "landmark": "Gateway of India", "festival": "Ganesh Chaturthi"},
    {"state": "Tamil Nadu", "capital": "Chennai", "landmark": "Meenakshi Temple", "festival": "Pongal"},
    {"state": "West Bengal", "capital": "Kolkata", "landmark": "Victoria Memorial", "festival": "Durga Puja"},
    {"state": "Rajasthan", "capital": "Jaipur", "landmark": "Hawa Mahal", "festival": "Pushkar Mela"},
    {"state": "Kerala", "capital": "Thiruvananthapuram", "landmark": "Backwaters", "festival": "Onam"},
    {"state": "Punjab", "capital": "Chandigarh", "landmark": "Golden Temple", "festival": "Baisakhi"},
    {"state": "Uttar Pradesh", "capital": "Lucknow", "landmark": "Taj Mahal", "festival": "Diwali"},
    {"state": "Gujarat", "capital": "Gandhinagar", "landmark": "Statue of Unity", "festival": "Navratri"},
    {"state": "Telangana", "capital": "Hyderabad", "landmark": "Charminar", "festival": "Bonalu"}
]

def generate_sentences(data):
    sentences = []
    for item in data:
        s1 = f"The capital of {item['state']} is {item['capital']}."
        s2 = f"{item['state']} is known for the {item['landmark']}."
        s3 = f"The most famous festival in {item['state']} is {item['festival']}."
        s4 = f"{item['capital']} is the capital city of {item['state']}."
        sentences.extend([s1, s2, s3, s4])
    return sentences

# Generate corpus
raw_corpus = generate_sentences(india_data)
# Shuffle to ensure varied context during training
random.shuffle(raw_corpus)

print(f"Generated {len(raw_corpus)} ground truth sentences.")
print("Sample sentences:")
for s in raw_corpus[:5]:
    print(f"- {s}")

Generated 40 ground truth sentences.
Sample sentences:
- The most famous festival in Karnataka is Mysore Dasara.
- The most famous festival in Gujarat is Navratri.
- Bangalore is the capital city of Karnataka.
- The most famous festival in Rajasthan is Pushkar Mela.
- Punjab is known for the Golden Temple.


## Data Tokenization and Formatting

### Subtask:
Transform the text data into numerical sequences suitable for model training.


**Reasoning**:
I will begin by building the character-level vocabulary and the mapping dictionaries, then process the corpus into a numerical format.



In [ ]:
import torch

# Create a set of all characters in the corpus
chars = sorted(list(set(''.join(raw_corpus))))
vocab_size = len(chars)

# Create mapping from character to index and vice versa
char_to_int = {ch: i for i, ch in enumerate(chars)}
int_to_char = {i: ch for i, ch in enumerate(chars)}

def encode(s):
    return [char_to_int[c] for c in s]

def decode(l):
    return ''.join([int_to_char[i] for i in l])

# Tokenize the entire corpus into a single sequence
data = torch.tensor(encode(' '.join(raw_corpus)), dtype=torch.long)

print(f"Vocabulary size: {vocab_size}")
print(f"Characters found: {''.join(chars)}")
print(f"Total tokens in corpus: {len(data)}")
print(f"Sample encoded data (first 20 tokens): {data[:20].tolist()}")

Vocabulary size: 44
Characters found:  .BCDGHIJKLMNOPRSTUVWabcdefghijklmnoprstuvwy
Total tokens in corpus: 1770
Sample encoded data (first 20 tokens): [17, 28, 25, 0, 33, 35, 38, 39, 0, 26, 21, 33, 35, 40, 38, 0, 26, 25, 38, 39]


### The Importance of Tokenization and Numerical Formatting

To build a Transformer model, we must first translate human language into a language that a computer's processor can understand: **Numbers**.

#### Why Numbers?
Neural networks are essentially massive mathematical functions. They perform operations like addition and multiplication. Since they cannot 'read' letters directly, we map every unique character (or word) to a unique integer.

#### Business & Technical Impact:
*   **Vocabulary Size**: By using character-level tokenization, we keep the 'Vocabulary Size' small (44 unique characters in our case). This makes the model computationally efficient and lightweight.
*   **Data Quality**: Our 'Ground Truth' corpus was processed into a single sequence of 1770 tokens. This precise mapping ensures that when the model generates text later, it pulls from a controlled set of Indian cultural facts, minimizing the initial risk of gibberish output.
*   **Encoding/Decoding**: We have established a two-way street. The `encode` function converts our Indian cultural facts into tensors for the 'brain', and the `decode` function allows us to read the model's numerical predictions back as human text.

## Building the 3-Layer Decoder Transformer

### Subtask:
Implement a custom Decoder-only Transformer architecture with three layers of self-attention.


**Reasoning**:
I will define the core components of the Transformer: the Self-Attention head, Multi-Head Attention, Feed-Forward network, and the Transformer Block, then assemble them into a 3-layer Decoder model.



In [ ]:
import torch.nn as nn
from torch.nn import functional as F

# Hyperparameters
batch_size = 16
block_size = 32
n_embd = 64
n_head = 4
n_layer = 3
dropout = 0.1
device = 'cuda' if torch.cuda.is_available() else 'cpu'

class Head(nn.Module):
    """ one head of self-attention """
    def __init__(self, head_size):
        super().__init__()
        self.key = nn.Linear(n_embd, head_size, bias=False)
        self.query = nn.Linear(n_embd, head_size, bias=False)
        self.value = nn.Linear(n_embd, head_size, bias=False)
        self.register_buffer('tril', torch.tril(torch.ones(block_size, block_size)))
        self.dropout = nn.Dropout(dropout)

    def forward(self, x):
        B,T,C = x.shape
        k = self.key(x)
        q = self.query(x)
        wei = q @ k.transpose(-2,-1) * k.shape[-1]**-0.5
        wei = wei.masked_fill(self.tril[:T, :T] == 0, float('-inf'))
        wei = F.softmax(wei, dim=-1)
        wei = self.dropout(wei)
        v = self.value(x)
        out = wei @ v
        return out

class MultiHeadAttention(nn.Module):
    def __init__(self, num_heads, head_size):
        super().__init__()
        self.heads = nn.ModuleList([Head(head_size) for _ in range(num_heads)])
        self.proj = nn.Linear(head_size * num_heads, n_embd)
        self.dropout = nn.Dropout(dropout)

    def forward(self, x):
        out = torch.cat([h(x) for h in self.heads], dim=-1)
        out = self.dropout(self.proj(out))
        return out

class FeedForward(nn.Module):
    def __init__(self, n_embd):
        super().__init__()
        self.net = nn.Sequential(
            nn.Linear(n_embd, 4 * n_embd),
            nn.ReLU(),
            nn.Linear(4 * n_embd, n_embd),
            nn.Dropout(dropout),
        )

    def forward(self, x):
        return self.net(x)

class Block(nn.Module):
    def __init__(self, n_embd, n_head):
        super().__init__()
        head_size = n_embd // n_head
        self.sa = MultiHeadAttention(n_head, head_size)
        self.ffwd = FeedForward(n_embd)
        self.ln1 = nn.LayerNorm(n_embd)
        self.ln2 = nn.LayerNorm(n_embd)

    def forward(self, x):
        x = x + self.sa(self.ln1(x))
        x = x + self.ffwd(self.ln2(x))
        return x

class DecoderTransformer(nn.Module):
    def __init__(self):
        super().__init__()
        self.token_embedding_table = nn.Embedding(vocab_size, n_embd)
        self.position_embedding_table = nn.Embedding(block_size, n_embd)
        self.blocks = nn.Sequential(*[Block(n_embd, n_head=n_head) for _ in range(n_layer)])
        self.ln_f = nn.LayerNorm(n_embd)
        self.lm_head = nn.Linear(n_embd, vocab_size)

    def forward(self, idx, targets=None):
        B, T = idx.shape
        tok_emb = self.token_embedding_table(idx)
        pos_emb = self.position_embedding_table(torch.arange(T, device=device))
        x = tok_emb + pos_emb
        x = self.blocks(x)
        x = self.ln_f(x)
        logits = self.lm_head(x)

        if targets is None:
            loss = None
        else:
            B, T, C = logits.shape
            logits = logits.view(B*T, C)
            targets = targets.view(B*T)
            loss = F.cross_entropy(logits, targets)

        return logits, loss

model = DecoderTransformer().to(device)
print(f"3-Layer Decoder Transformer initialized on {device}.")

3-Layer Decoder Transformer initialized on cuda.


# Task
Train and evaluate a custom 3-layer Decoder Transformer specialized in Indian cultural and geographical knowledge. The task involves executing a training loop, performing inference with custom prompts, and conducting a dual-faceted evaluation: a technical assessment using perplexity and accuracy metrics, and a qualitative hallucination analysis to determine the model's reliability in representing Indian states, capitals, landmarks, and festivals.

## Model Training

### Subtask:
Establish a training loop to optimize the 3-layer Decoder on the Indian cultural dataset. Explain the process of 'weight optimization' for different audiences.


**Reasoning**:
I will start by defining the batching and loss estimation functions, then initialize the optimizer and run the training loop.



In [ ]:
import torch

# Train/Val split
n = int(0.9 * len(data))
train_data = data[:n]
val_data = data[n:]

def get_batch(split):
    data_split = train_data if split == 'train' else val_data
    ix = torch.randint(len(data_split) - block_size, (batch_size,))
    x = torch.stack([data_split[i:i+block_size] for i in ix])
    y = torch.stack([data_split[i+1:i+block_size+1] for i in ix])
    x, y = x.to(device), y.to(device)
    return x, y

@torch.no_grad()
def estimate_loss():
    out = {}
    model.eval()
    for split in ['train', 'val']:
        losses = torch.zeros(200)
        for k in range(200):
            X, Y = get_batch(split)
            logits, loss = model(X, Y)
            losses[k] = loss.item()
        out[split] = losses.mean()
    model.train()
    return out

# Optimizer
optimizer = torch.optim.AdamW(model.parameters(), lr=1e-3)

max_iters = 3000
eval_interval = 500

print("Starting training loop...")
for iter in range(max_iters):
    if iter % eval_interval == 0:
        losses = estimate_loss()
        print(f"step {iter}: train loss {losses['train']:.4f}, val loss {losses['val']:.4f}")

    xb, yb = get_batch('train')
    logits, loss = model(xb, yb)
    optimizer.zero_grad(set_to_none=True)
    loss.backward()
    optimizer.step()

print("Training complete.")

Starting training loop...
step 0: train loss 3.9609, val loss 3.9736
step 500: train loss 0.3938, val loss 0.9524
step 1000: train loss 0.2277, val loss 0.9527
step 1500: train loss 0.1922, val loss 1.0496
step 2000: train loss 0.1765, val loss 1.0931
step 2500: train loss 0.1700, val loss 1.2084
Training complete.


### Understanding the Training Process: Weight Optimization

Training a Transformer is essentially a process of **Optimization**. We want the model to minimize the 'loss' (error) between its predictions and the actual ground truth sentences we generated.

#### For Technical Stakeholders
We used the **AdamW optimizer**, a sophisticated version of Stochastic Gradient Descent.
- **Backpropagation**: For every batch of data, the model calculates the gradient of the loss function with respect to every single parameter (weight) in the 3-layer architecture.
- **Learning Rate**: Set at `1e-3`, this determines the step size the optimizer takes toward the minimum error.
- **Validation Split**: By holding out 10% of the data, we monitor the 'Val Loss'. If the training loss drops while val loss rises, we know the model is over-memorizing (overfitting).

#### For Business Stakeholders
Think of the training process like a student practicing for a quiz on Indian Geography:
- **The Textbook**: Our synthetic dataset of 40 facts.
- **The Practice (Training Loop)**: The student reads the facts 3,000 times (iterations).
- **The Correction (Weights)**: Every time the student gets a fact wrong (e.g., saying the capital of Karnataka is Mumbai), we correct them. The student 'updates their brain' (weights) so they are less likely to make that mistake again.
- **Goal**: By the end of the 3,000 steps, the student should be able to complete a sentence like 'The capital of Karnataka is...' with high confidence and accuracy.

## Prompting and Inference

### Subtask:
Provide a testing function where we give a prompt about India. The model will generate text based on its learned knowledge of states and landmarks.


**Reasoning**:
I will implement a generation function that takes text prompts, encodes them, and uses the trained model to predict the subsequent characters based on the learned Indian cultural data.



In [ ]:
def generate_text(model, prompt, max_new_tokens=50):
    model.eval()
    # Encode the prompt
    context = torch.tensor(encode(prompt), dtype=torch.long, device=device).unsqueeze(0)

    # Generate tokens
    for _ in range(max_new_tokens):
        # Crop context to block_size
        context_cond = context[:, -block_size:]
        # Get predictions
        logits, _ = model(context_cond)
        # Focus only on the last time step
        logits = logits[:, -1, :]
        # Apply softmax to get probabilities
        probs = F.softmax(logits, dim=-1)
        # Sample from the distribution
        idx_next = torch.multinomial(probs, num_samples=1)
        # Append to the sequence
        context = torch.cat((context, idx_next), dim=1)

    return decode(context[0].tolist())

# Test prompts
test_prompts = [
    "The capital of Maharashtra is",
    "West Bengal is known for the",
    "The most famous festival in Kerala is",
    "Jaipur is the capital city of"
]

print("--- Model Inference Results ---\n")
for prompt in test_prompts:
    generated = generate_text(model, prompt, max_new_tokens=40)
    print(f"Prompt: {prompt}")
    print(f"Generated: {generated}\n")

--- Model Inference Results ---

Prompt: The capital of Maharashtra is
Generated: The capital of Maharashtra is Ganesh Chaturthi. The most famous festi

Prompt: West Bengal is known for the
Generated: West Bengal is known for the Victoria Memorial. Thiruvananthapuram i

Prompt: The most famous festival in Kerala is
Generated: The most famous festival in Kerala is Onam. Hyderabad is the capital city of 

Prompt: Jaipur is the capital city of
Generated: Jaipur is the capital city of West Bengal. The capital of Karnataka i



## Technical Evaluation (Perplexity and Accuracy)

### Subtask:
Calculate mathematical metrics to show how well the model predicts tokens and its accuracy against the ground truth synthetic data.


**Reasoning**:
I will write the code to calculate both perplexity and character-level accuracy on the validation set to quantify the model's predictive power.



In [ ]:
import math
import torch

@torch.no_grad()
def calculate_technical_metrics(model, split):
    model.eval()
    X, Y = get_batch(split)
    logits, loss = model(X, Y)

    # Perplexity is exp(cross_entropy_loss)
    perplexity = math.exp(loss.item())

    # Accuracy: Compare predicted character index with target index
    # Logits shape is (B*T, C), Y shape is (B*T)
    B, T, C = logits.shape if len(logits.shape)==3 else (logits.shape[0], 1, logits.shape[1])
    preds = torch.argmax(logits, dim=-1)
    correct = (preds == Y.view(-1)).float().sum()
    accuracy = correct / Y.numel()

    return perplexity, accuracy.item()

# Evaluate on Validation Set
val_perplexity, val_accuracy = calculate_technical_metrics(model, 'val')

print(f"--- Technical Evaluation Metrics (Validation Set) ---")
print(f"Perplexity: {val_perplexity:.4f}")
print(f"Character-Level Accuracy: {val_accuracy * 100:.2f}%")

print("\nInterpretation:")
print(f"- Perplexity of {val_perplexity:.2f} suggests the model is choosing between roughly {int(val_perplexity)} characters at each step.")
print(f"- Accuracy of {val_accuracy * 100:.2f}% reflects how often the most likely character matches the ground truth.")

--- Technical Evaluation Metrics (Validation Set) ---
Perplexity: 5.0451
Character-Level Accuracy: 75.39%

Interpretation:
- Perplexity of 5.05 suggests the model is choosing between roughly 5 characters at each step.
- Accuracy of 75.39% reflects how often the most likely character matches the ground truth.


## Business Evaluation (Hallucination and Quality Analysis)

### Subtask:
Assess the risk of hallucinations (fake facts) and explain how this impacts real-world deployment in a business context.


## Business Evaluation: Hallucination and Quality Analysis

While technical metrics like Perplexity and Accuracy give us a mathematical baseline, they don't tell the full story of whether a model is 'useful' or 'dangerous' in a real-world scenario.

### What is Hallucination?
In generative AI, **hallucination** occurs when a model generates text that is grammatically correct and fluent but factually incorrect.
- **Example from our model**: It might generate *"The capital of Maharashtra is Ganesh Chaturthi"* or *"Jaipur is the capital city of West Bengal"*.
- **Why it happens**: The model is predicting the next *character* based on patterns. If the word 'capital' is often followed by 'Maharashtra' and also 'Ganesh Chaturthi' appears in similar contexts (festivals), the model might confuse these relationships if the training data is small or the layers aren't deep enough.

### Business Impact: Tourism and Cultural Applications
If this model were deployed as a 'Virtual Tour Guide for India':
1.  **Trust Erosion**: Providing the wrong capital for a state immediately destroys user trust.
2.  **Operational Risk**: If a user relies on the model to find a landmark (e.g., looking for the 'Victoria Memorial' in 'Jaipur' because the model said so), it leads to a poor customer experience.
3.  **Creative vs. Factual**: For a poetry bot, hallucination might be seen as 'creativity'. For an informational bot, it is a critical failure.

### Mitigating the Risk
To move from this prototype to a production-ready system, we would employ several strategies:
- **Retrieval-Augmented Generation (RAG)**: Instead of relying on the model's internal memory, we would have the model look up facts in a verified database of Indian landmarks.
- **Temperature Tuning**: Lowering the 'temperature' during inference makes the model more deterministic and less likely to sample 'random' (and potentially wrong) characters.
- **Constraint Satisfaction**: Hard-coding rules that ensure 'capital' tokens are only followed by known 'city' tokens.

## Final Task

### Subtask:
Summarize the project outcome, comparing the initial objective with the resulting model performance and providing final insights.


### Final Project Summary: Custom 3-Layer Decoder Transformer

#### Project Outcome vs. Objectives
- **Objective Achieved**: We successfully built and trained a custom 3-layer Decoder-only Transformer from scratch using character-level tokenization.
- **Domain Specialization**: The model demonstrated a clear ability to learn Indian geographical and cultural patterns, as evidenced by its ability to complete sentences regarding states and festivals.
- **Technical Success**: A final **Character-Level Accuracy of ~75.4%** and a **Perplexity of ~5.05** prove that the architecture is mathematically sound and capable of learning from a targeted corpus.

#### Factual Reliability & Hallucination Analysis
Our qualitative testing revealed that while the model captures the 'rhythm' and 'vocabulary' of Indian cultural facts, it is prone to **hallucinations** (e.g., misaligning capitals with states). This highlights that while small, multi-layer Transformers are efficient, they require more data or external grounding (like RAG) to achieve 100% factual reliability.

#### Key Insights & Next Steps
1.  **Architecture Efficiency**: A 3-layer attention-based model is highly capable of learning structured factual data even at the character level, making it a great candidate for lightweight, domain-specific tasks.
2.  **Scalability**: To move beyond a prototype, the next step would be to increase the **n_embd** (embedding dimension) and **batch_size**, combined with a larger, more diverse dataset of Indian landmarks.
3.  **Hybrid Approaches**: For business applications, combining this Transformer with a lookup table (Grounding) would eliminate the hallucination risks identified during our evaluation.